auto3dseg/algorithm_templates/segresnet/scripts/segmenter.py

```diff
@@ -0,0 +1,1395 @@
+                    wrtr.writerow(header)
+            if len(kwargs):
+                with open(csv_path, "a") as myfile:
+                    wrtr = csv.writer(myfile, delimiter="\t")
+                    wrtr.writerow(list(kwargs.values()))
+
+    def save_progress_yaml(self, progress_path=None, ckpt=None, **report):
+
+        if ckpt is not None:
+            report["model"] = ckpt
+
+        report["date"] = str(datetime.now())[:19]
+
+        if progress_path is not None:
+            yaml.add_representer(
+                float, lambda dumper, value: dumper.represent_scalar("tag:yaml.org,2002:float", "{0:.4f}".format(value))
+            )
+            with open(progress_path, "a") as progress_file:
+                yaml.dump([report], stream=progress_file, allow_unicode=True, default_flow_style=None, sort_keys=False)
+
+        print("Progress:", ",".join(f" {k}: {v}" for k, v in report.items()))
+
+    def run(self):
+        if self.config["validate"]["enabled"]:
+            self.validate()
+        elif self.config["infer"]["enabled"]:
+            self.infer()
+        else:
+            self.train()
+
+
+def run_segmenter_worker(rank=0, config_file: Optional[Union[str, Sequence[str]]] = None, override: Dict = {}):
+
+    logging.basicConfig(stream=sys.stdout, level=logging.INFO)
+    dist_available = dist.is_available()
+
+    print(rank, "run_segmenter_worker dist_available:", dist_available, "initialized", dist.is_initialized())
+
+    if dist_available:
+        mgpu = override.get("mgpu", None)
+        if mgpu is not None:
+            dist.init_process_group(backend="nccl", rank=rank, **mgpu)  # we spawn this process
+            mgpu["rank"] = rank
+            if rank == 0:
+                print("Distributed: initializing multi-gpu tcp:// process group", mgpu)
+
+        elif dist.is_torchelastic_launched():
+
+            rank = int(os.getenv("LOCAL_RANK"))
+            world_size = int(os.getenv("LOCAL_WORLD_SIZE"))
+            dist.init_process_group(backend="nccl", init_method="env://")  # torchrun spawned it
+            override["mgpu"] = {"world_size": world_size, "rank": rank}
+            print("Distributed: initializing multi-gpu env:// process group", override["mgpu"])
+
+    segmenter = Segmenter(config_file=config_file, config_dict=override, rank=rank)
+    best_metric = segmenter.run()
+    segmenter = None
+
+    if dist_available and dist.is_initialized():
+        dist.destroy_process_group()
+
+    return best_metric
+
+
+def run_segmenter(config_file: Optional[Union[str, Sequence[str]]] = None, **kwargs):
+    """
+    if multiple gpu available, start multiprocessing for all gpus
+    """
+    nprocs = torch.cuda.device_count()
+    if nprocs > 1 and not dist.is_torchelastic_launched():
+        kwargs["mgpu"] = {"world_size": nprocs, "init_method": kwargs.get("init_method", "tcp://127.0.0.1:23456")}
+        torch.multiprocessing.spawn(run_segmenter_worker, nprocs=nprocs, args=(config_file, kwargs))
```

Comment:
Hi @Nic-Ma , I remembered you talked about some overhead issue of using `spawn` in some special cases. Can you confirm it won't affect here? Thanks!

Reply:
the code supports both torchrun .. and direct distributed spawning.  The lines you highlighted will spawn manually (if torchrun wasn't used). But if a user uses torchrun ..., those lines won't be called.  torchrun also uses spawn(). so we actually always use spawn() in distributed training.